### PR TITLE
[move-lang] Added vector literals

### PR DESCRIPTION
**Pull request by tnowacki**
- Added vector literals
- Vector literals get compiled to constants when they can be
- Otherwise, they compile to the new `VecPack` instruction 

- Currently the syntax is straight-forward, with just a tad oddity of an edge case
  - The syntax is `vector<t>[e0, e1, ..., en]` where the type argument is optional. `vector<t>[]` if it is empty (type argument still optional)
  - The odd edge case is that I didn't make `vector` a keyword. So it still parses as an identifier except in the following situation: A term followed by a `[` or `<` with no space. So `vector <` will parse `vector` as an identifier. 
  - We could just make `vector` a keyword when it is a term. This isn't a big deal since we restrict the usage of the identifier after parsing. 
  - In other words, it basically is a keyword, but not in the parser for simplicity 

## Motivation

- Literals are neat
- We need a good way to compile to the new `VecPack` instruction 

## Test Plan

- Added new tests

### CODE REVIEW

**Comment by bob-wilson** [comment]
Is this something that you're trying to get in for 1.5? Otherwise, I suggest we hold off landing until after the branch. (I haven't reviewed the content yet.)

File: language/move-lang/src/parser/syntax.rs [hunk 1]
```diff
@@ -964,19 +1009,22 @@ fn parse_name_exp(context: &mut Context) -> Result<Exp_, Diagnostic> {
     let n = parse_name_access_chain(context, || {
         panic!("parse_name_exp with something other than a ModuleAccess")
     })?;
+    parse_name_exp_after_access_chain(context, n)
+}
 
+fn parse_name_exp_after_access_chain(
```
Comment (bob-wilson) [inline language/move-lang/src/parser/syntax.rs:1015]: I don't see this `parse_name_exp_after_access_chain` function used elsewhere. Am I missing something or did you split it just to keep the functions smaller?

File: language/move-lang/tests/move_check/typing/constant_non_base_type.exp [hunk 2]
```diff
@@ -55,9 +55,3 @@ error[E04012]: invalid type for constant
   │               Unpermitted constant type
   │               Found: '(_, _)'. But expected one of: 'u8', 'u64', 'u128', 'bool', 'address', 'vector<_>'
 
-error[E04013]: invalid statement or expression in constant
```
Comment (bob-wilson) [inline language/move-lang/tests/move_check/typing/constant_non_base_type.exp:58]: Why did this stop being an error? If this is intentional, it seems like it should be in a separate test instead of being grouped with 3 of things that are expected to fail.

Comment (bob-wilson) [reply language/move-lang/tests/move_check/typing/constant_non_base_type.exp:58]: Ah, I see there was a test case added to `constant_supported_exps.move`. I think it should be removed from here, though.

**Review by bob-wilson — COMMENTED** [review]
I left a few comments. I also noticed that this change adds support for expression lists in constants. I wasn't expecting that. Be sure to update the docs for that as well.

File: language/move-lang/src/parser/syntax.rs (hunk 1 above)
Comment (tnowacki) [reply language/move-lang/src/parser/syntax.rs:1015]: I think I split it for an older version of the parser and didn't unsplit it

File: language/move-lang/tests/move_check/typing/constant_non_base_type.exp (hunk 2 above)
Comment (tnowacki) [reply language/move-lang/tests/move_check/typing/constant_non_base_type.exp:58]: well... maybe. The test still needs to see that you can't do `const C: (u64, u64) = (0, 1);` That type is still disallowed, but the expression is now allowed. There is no way of really using it though outside of letbindings, so I doubt anyone will even notice/care

Comment (bob-wilson) [reply language/move-lang/tests/move_check/typing/constant_non_base_type.exp:58]: Got it. I had missed that the previous error is for this same line of code.

File: language/move-lang/src/parser/syntax.rs [hunk 3]
```diff
@@ -874,6 +882,43 @@ fn parse_term(context: &mut Context) -> Result<Exp, Diagnostic> {
             Exp_::Continue
         }
 
+        Tok::IdentifierValue if context.tokens.content() == VECTOR_IDENT => {
```
Comment (tnowacki) [inline language/move-lang/src/parser/syntax.rs:885]: @bob-wilson thoughts on this hack?

Particularly on the casing of the next token being `<` or `[`. 
Should we just make `vector` a conditional keyword? i.e. a keyword when in an exp context? It is already restricted from being used as a name for anything (anything other than the type) so it won't change much

Comment (bob-wilson) [reply language/move-lang/src/parser/syntax.rs:885]: I don't know what "keyword when in an exp context" would mean. The lexer doesn't (currently) distinguish the context, so we would need to add that extra complexity. Maybe that would still be cleaner and/or more maintainable this this hack? From the user's point of view, would it change anything? I don't have much of an opinion offhand. Maybe @modocache has something to say about the potential impact on regexps for syntax coloring?

Comment (tnowacki) [reply language/move-lang/src/parser/syntax.rs:885]: There is no extra complexity. We already conditional keyword all over the place, you just case out the content of the identifier before allowing the parser to continue. Which is basically what has happened here except it reverts back to a name if the next token isn't `<` or `\s*[`

Comment (tnowacki) [reply language/move-lang/src/parser/syntax.rs:886]: @bob-wilson I added a fixup that deals just with this issue. With the fixup, `vector` is _not_ a conditional keyword in expressions. This has a bunch of downstream effects you can see in tests

File: language/move-lang/src/expansion/translate.rs [hunk 4]
```diff
@@ -2480,11 +2488,14 @@ pub fn is_valid_struct_constant_or_schema_name(s: &str) -> bool {
     s.starts_with(|c| matches!(c, 'A'..='Z'))
 }
 
-fn check_restricted_self_name(context: &mut Context, case: &str, n: &Name) -> Result<(), ()> {
-    if n.value.as_str() == ModuleName::SELF_NAME {
+// Checks for a restricted name in any decl case
+// Self and vector are nto allowed
```
Comment (bob-wilson) [inline language/move-lang/src/expansion/translate.rs:2492]:
```suggestion
// Self and vector are not allowed
```

File: language/move-lang/src/parser/syntax.rs (hunk 3 above)
Comment (bob-wilson) [reply language/move-lang/src/parser/syntax.rs:886]: This version is simpler and disallowing "vector" for variable/module names seems like a good idea anyway to avoid confusion, so this looks good to me.

**Review by bob-wilson — DISMISSED** [review]

File: language/documentation/book/src/vector.md [hunk 5]
```diff
@@ -6,6 +6,86 @@ collection of `T`'s that can grow or shrink by pushing/popping values off the "e
...
+## Literals
+
+### General `vector` Literals
+
+Vectors of any type can be created with `vector` literals.
+
+| Syntax                | Type                                                                           | Description                                |
+| --------------------- | ------------------------------------------------------------------------------ | ------------------------------------------ |
+| `vector[]`            | `vector[]: vector<T>` where `T` is any single, non-reference type              | An empty vector                            |
+| `vector[e1, ..., en]` | `vector[e1, ..., en]: vector<T>` where `e_i: T` s.t. `0 <= i <= n` and `n > 0` | A vector with `n` elements (of length `n`) |
```
Comment (bob-wilson) [inline language/documentation/book/src/vector.md:18]: Should that be "0 < i <= n" (instead of less or equal)?

File: language/documentation/book/src/vector.md [hunk 6]
```diff
@@ -6,6 +6,86 @@ collection of `T`'s that can grow or shrink by pushing/popping values off the "e
...
+
+In these cases, the type of the `vector` ins inferred, either from the element type or from the
```
Comment (bob-wilson) [inline language/documentation/book/src/vector.md:20]:
```suggestion
In these cases, the type of the `vector` is inferred, either from the element type or from the
```

File: language/documentation/book/src/vector.md [hunk 7]
```diff
@@ -6,6 +6,86 @@ collection of `T`'s that can grow or shrink by pushing/popping values off the "e
...
+vectors usage. If the type cannot be inferred, or simply for added clarity, the type can be specfied
```
Comment (bob-wilson) [inline language/documentation/book/src/vector.md:21]:
```suggestion
vector's usage. If the type cannot be inferred, or simply for added clarity, the type can be specified
```

**Review by bob-wilson — CHANGES_REQUESTED** [review]
Some small comments on the docs

File: language/move-lang/tests/move_check/expansion/restricted_address_names.exp [hunk 8]
```diff
@@ -1,12 +1,12 @@
-error[E02010]: invalid name
+error[E03011]: invalid use of reserved name
   ┌─ tests/move_check/expansion/restricted_address_names.move:1:8
   │
 1 │ module Self::A {}
-  │        ^^^^ Invalid named address 'Self'. 'Self' is restricted and cannot be used to name an address
+  │        ^^^^ Invalid address name 'Self'. 'Self' is restricted and cannot be used to name a address
```
Comment (bob-wilson) [inline language/move-lang/tests/move_check/expansion/restricted_address_names.exp:5]: This should be "an address"

Comment (tnowacki) [reply language/move-lang/tests/move_check/expansion/restricted_address_names.exp:5]: oof testing my rust skills let's see how this works lol

**Review by tzakian — APPROVED** [review]

**Review by tzakian — APPROVED** [review]
